File: Homework/quoc_khanh/lesson_1/206.ReverseLinkedList.py
# Definition for singly-linked list.
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def reverseInRecursion(self, prev: ListNode, curr: ListNode) -> ListNode:
        if curr == None:
            return prev
        next = curr.next
        curr.next = prev
        return self.reverseInRecursion(curr, next)
    
    def reverse(self, head: ListNode) -> ListNode:
        if head.next == None:
            return head
        newHead = head
        currNode = head.next
        head.next = None
        while currNode != None:
            nextNode = currNode.next
            currNode.next = newHead
            newHead = currNode
            currNode = nextNode
        return newHead

    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head == None:
            return None
        # return self.reverseInRecursiov(None, head)
        return self.reverse(head)

File: Homework/Hoang_Nguyen/lesson 1/206_reverse_linked_list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var reverseList = function (head) {
  let prev = null;
  let cur = head;
  while (cur) {
    let next = cur.next;
    cur.next = prev;

    prev = cur;
    cur = next;
  }
  return prev;
};

/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var reverseListRecur = function (head) {
  if (!head) return head;
  if (!head.next) return head;
  let result = reverseListRecur(head.next);
  head.next.next = head;
  head.next = null;
  return result;
};
